llvm/include/llvm/ProfileData/SampleProf.h

```diff
@@ -975,6 +975,8 @@ class FunctionSamples {
     return CallsiteSamples;
   }
 
+  CallsiteSampleMap &getCallsiteSamples() { return CallsiteSamples; }
```

Comment:
I remember we intentionally not exposing this as non-const (it was added once, and later cleaned up and removed). In generally we should be careful about exposing non-const accessor to private members. 

OTOH, I'm wondering do we actually need to filter call site (inlined) samples? I think only top level functions are indistinguishable, and inlinee's samples can be distinguished and still used by their inliner?

Also __cxx_global_var_init can inlined only into tls_init, but tls_ininit probably will never be inlined anywhere? 

If we really need to do this, it's probably better to use the const accessor and then cast away the constness, rather than adding an non-const API which is essentially saying it's expected to be modified in general and it's likely subject to misuse in the future.

Reply:
yeah, we intentionally removed this.. I got it, will remove this.

>OTOH, I'm wondering do we actually need to filter call site (inlined) samples? 
>Also __cxx_global_var_init can inlined only into tls_init, but tls_ininit probably will never be inlined anywhere?

I do see `tls_init` appearing in the `CallsiteSamples`, presumably it can be inlined. 
Another thing which is specific to profile matching, we use the flattened profile for profile matching, so the inlinee's samples will be merged into top-level function, then used as the top-level function in profile matching(causing the issues we hit)

> I think only top level functions are indistinguishable, and inlinee's samples can be distinguished and still used by their inliner?
I think you are right, so we can leave it only for top level function if we can ensure it's also fine for profile matching.

Reply:
OK, I just checked the profile matching, so we only check against `FSFlattened` (not FS) https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/SampleProfile.cpp#L2433
The reason is that there could be function that all the call are inlined, so there is no top-level version, but it's still can be checksum mismatched, we should make it possible to run stale matching for those functions.

Given this, we probably needs the filtering for inlined samples.